pkg/cloudprovider/providers/azure/azure_storage.go

```diff
@@ -0,0 +1,137 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azure
+
+import (
+	"github.com/Azure/azure-sdk-for-go/arm/compute"
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+)
+
+// attach a vhd to vm
+// the vhd must exist, can be identified by diskName, diskUri, and lun.
+func (az *Cloud) AttachDisk(diskName, diskUri, vmName string, lun int32, cachingMode compute.CachingTypes) error {
+	vm, exists, err := az.getVirtualMachine(vmName)
+	if err != nil {
+		return err
+	} else if !exists {
+		return cloudprovider.InstanceNotFound
+	}
+	disks := *vm.Properties.StorageProfile.DataDisks
+	disks = append(disks,
+		compute.DataDisk{
+			Name: &diskName,
+			Vhd: &compute.VirtualHardDisk{
+				URI: &diskUri,
+			},
+			Lun:          &lun,
+			Caching:      cachingMode,
+			CreateOption: "attach",
+		})
+
+	newVM := compute.VirtualMachine{
+		Location: vm.Location,
+		Properties: &compute.VirtualMachineProperties{
+			StorageProfile: &compute.StorageProfile{
+				DataDisks: &disks,
+			},
+		},
+	}
+	res, err := az.VirtualMachinesClient.CreateOrUpdate(az.ResourceGroup, vmName,
+		newVM, nil)
+	glog.V(4).Infof("azure attach result:%#v, err: %v", res, err)
+	return err
+}
+
+// detach a vhd from host
+// the vhd can be identified by diskName or diskUri
+func (az *Cloud) DetachDiskByName(diskName, diskUri, vmName string) error {
+	vm, exists, err := az.getVirtualMachine(vmName)
+	if err != nil {
+		return err
+	} else if !exists {
+		return cloudprovider.InstanceNotFound
+	}
+	disks := *vm.Properties.StorageProfile.DataDisks
+	for i, disk := range disks {
+		if (disk.Name != nil && diskName != "" && *disk.Name == diskName) ||
+			(disk.Vhd.URI != nil && diskUri != "" && *disk.Vhd.URI == diskUri) {
+			// found the disk
+			glog.V(4).Infof("detach disk: lun %d name %q uri %q", *disk.Lun, diskName, diskUri)
+			disks = append(disks[:i], disks[i+1:]...)
+			break
+		}
+	}
+	newVM := compute.VirtualMachine{
+		Location: vm.Location,
+		Properties: &compute.VirtualMachineProperties{
+			StorageProfile: &compute.StorageProfile{
+				DataDisks: &disks,
+			},
+		},
+	}
+	res, err := az.VirtualMachinesClient.CreateOrUpdate(az.ResourceGroup, vmName,
+		newVM, nil)
+	glog.V(4).Infof("azure detach result:%#v, err: %v", res, err)
+	return err
+}
+
+// given a vhd's diskName and diskUri, find the lun on the host that the vhd is attached to
+func (az *Cloud) GetDiskLun(diskName, diskUri, vmName string) (int32, error) {
+	vm, exists, err := az.getVirtualMachine(vmName)
+	if err != nil {
+		return -1, err
+	} else if !exists {
+		return -1, cloudprovider.InstanceNotFound
+	}
+	disks := *vm.Properties.StorageProfile.DataDisks
+	for _, disk := range disks {
+		if disk.Lun != nil {
```

Comment:
join these conditions together, there's no need for multiple `if` statements.  You can use line breaks if you want clarity about what does what.